8295661: CompileTask::compile_id() should be passed as int

dafedafe opened the pull request:
Changed return type of `CompileTask::compile_id()` from `int` to `uint`.
Also modified the type everywhere `compile_id` is used (parameters, local variables, string formatting, constant assignments) *as much as possible*.
Added *asserts* to check for valid value range where not possible.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295661](https://bugs.openjdk.org/browse/JDK-8295661): CompileTask::compile_id() should be passed as int


### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Doug Simon](https://openjdk.org/census#dnsimon) (@dougxc - Committer) ⚠️ Review applies to [22f0a9f6](https://git.openjdk.org/jdk/pull/11630/files/22f0a9f60fbcb178438cdc79c98d6214bb1eafff)
 * [Tom Rodriguez](https://openjdk.org/census#never) (@tkrodriguez - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11630/head:pull/11630` \
`$ git checkout pull/11630`

Update a local copy of the PR: \
`$ git checkout pull/11630` \
`$ git pull https://git.openjdk.org/jdk pull/11630/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11630`

View PR using the GUI difftool: \
`$ git pr show -t 11630`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11630.diff">https://git.openjdk.org/jdk/pull/11630.diff</a>

</details>

dafedafe commented:
/covered

dafedafe commented:
/label add hotspot-compiler

TobiHartmann commented on src/hotspot/share/jvmci/jvmciEnv.cpp:
```diff
@@ -1622,7 +1622,9 @@ nmethod* JVMCIEnv::lookup_nmethod(address code, jlong compile_id_snapshot) {
   if (cb == (CodeBlob*) code) {
     nmethod* nm = cb->as_nmethod_or_null();
-    if (nm != NULL && (compile_id_snapshot == 0 || nm->compile_id() == compile_id_snapshot)) {
+    assert(compile_id_snapshot >= 0, "negative compile id snapshot");
```
That's something to check with the JVMCI experts but I'm wondering why `compile_id_snapshot` is a jlong whereas the compile id is a jint.

dougxc replied:
It's a good question and unfortunately I cannot recall the reason. Maybe @tkrodriguez can. In any case, I don't think it impacts this change.

tkrodriguez replied:
That's far enough back that I don't really know why I used jlong.  Its type should really match the real compile id type whatever that is.  It's also fully internal to the JVMCI implementation so you can change it without affecting Graal.

TobiHartmann commented on src/hotspot/share/jvmci/jvmciRuntime.cpp:
```diff
@@ -2020,8 +2020,9 @@ void JVMCIRuntime::compile_method(JVMCIEnv* JVMCIENV, JVMCICompiler* compiler, c
+  assert(compile_state->task()->compile_id() <= INT_MAX, "compile id too big");
   JVMCIObject result_object = JVMCIENV->call_HotSpotJVMCIRuntime_compileMethod(receiver, jvmci_method, entry_bci,
-                                                                     (jlong) compile_state, compile_state->task()->compile_id());
+                                                                     (jlong) compile_state, (int) compile_state->task()->compile_id());
```
Should be a cast to `jint`, right?

```suggestion
                                                                     (jlong) compile_state, (jint) compile_state->task()->compile_id());
```

TobiHartmann commented on src/hotspot/share/c1/c1_Compilation.cpp:
```diff
@@ -613,7 +613,7 @@ Compilation::Compilation(AbstractCompiler* compiler, ciEnv* env, ciMethod* metho
 Compilation::~Compilation() {
   // simulate crash during compilation
-  assert(CICrashAt < 0 || (uintx)_env->compile_id() != (uintx)CICrashAt, "just as planned");
+  assert(CICrashAt >= UINT_MAX || _env->compile_id() != CICrashAt, "just as planned");
```
I'm wondering if we should remove the `CICrashAt >= UINT_MAX` part to catch an UINT overflow before it can happen?

```suggestion
  assert(_env->compile_id() != CICrashAt, "just as planned");
```

TobiHartmann commented on src/hotspot/share/jvmci/jvmciCompilerToVM.cpp:
```diff
@@ -1179,7 +1179,7 @@ C2V_VMENTRY_0(jboolean, setCountersSize, (JNIEnv* env, jobject, jint new_size))
 C2V_END
 
-C2V_VMENTRY_0(jint, allocateCompileId, (JNIEnv* env, jobject, ARGUMENT_PAIR(method), int entry_bci))
+C2V_VMENTRY_0(juint, allocateCompileId, (JNIEnv* env, jobject, ARGUMENT_PAIR(method), int entry_bci))
```
This is called from Java code which does not have an unsigned int. I think we need to leave this as `jint` here.

TobiHartmann reviewed (commented):
Looks good, I added a few comments. Someone more familiar with JVMCI should have a look as well.

dafedafe reviewed (commented):
Thanks a lot for the review. I should have addressed all your comments.

TobiHartmann commented:
Thanks for making these changes. Looks good!

dougxc reviewed (approved):
The JVMCI changes in this PR look correct. @tkrodriguez could also please look them over.

tkrodriguez commented:
Why isn't the proper fix to simply change the `CompileTask::_compile_id` to an int?  Every other piece of code thinks it's an int so we should that declaration be controlling?  `CompileBroker::assign_compile_id` returns int which seems much more definitive to me.  Is there some known problem with running out of compile ids in the in signed range?

If this range is really being promoted to unsigned int then I think this needs to be properly exposed all the way through the JVMCI API with some new API to expose it as a long.  Simply adding asserts for something we believe can legally occur really isn't sufficient.

dafedafe commented:
@tkrodriguez thanks a lot for your insight. I see your point: the pragmatic approach you propose is surely more appropriate. Reverting the change.

dafedafe commented:
@tkrodriguez @TobiHartmann @dougxc I've reverted the previous changes and changed `CompileTask::_compile_id` to an `int` instead. I've also changed the type of the compile id from `uint` to `int` in a few other places to make things consistent.

TobiHartmann reviewed (approved):
Using unsigned int consistently feels appropriate to represent a compilation id, especially since we are already mixing unsigned and signed integers. But given the impact on JVMCI and the leakage into Java code which does not support an unsigned int, it's reasonable to use int consistently instead. The updated changes look good to me.

TobiHartmann commented:
Thanks Tom. The intention of this change was mainly consistency, not avoiding a potential overflow.

dafedafe commented:
Thanks @tkrodriguez @TobiHartmann @dougxc for the reviews!

dafedafe commented:
@TobiHartmann @tkrodriguez @dougxc thanks for your reviews.

dafedafe commented:
/integrate

TobiHartmann commented:
/sponsor
